pandas/core/reshape/merge.py

```diff
@@ -2197,11 +2170,23 @@ def injection(obj: ArrayLike):
             else:
                 # We get here with non-ndarrays in test_merge_by_col_tz_aware
                 #  and test_merge_groupby_multiple_column_with_categorical_column
-                lbv = flip(left_by_values)
-                rbv = flip(right_by_values)
-                lbv = ensure_object(lbv)
-                rbv = ensure_object(rbv)
-
+                mapped = [
+                    _factorize_keys(
```

Comment:
I'm assuming this work alright with ExtensionArrays? Do we have tests for merging on `by` that is a extension column

Reply:
Yes, this will work with EAs. I've parameterized an existing test to include EA dtypes.

Reply:
Added an additional test as this seems to fix #43541